func.go

```diff
@@ -9,45 +9,65 @@ import (
 )
 
 type dslFunction struct {
-	name string
+	IsCacheable bool
+	Name        string
 	// if numberOfArgs is defined the signature is automatically generated
-	numberOfArgs       int
-	signatures         []string
-	expressionFunction govaluate.ExpressionFunction
+	NumberOfArgs       int
+	Signatures         []string
+	ExpressionFunction govaluate.ExpressionFunction
 }
 
-func (d dslFunction) Signatures() []string {
+func (d dslFunction) GetSignatures() []string {
 	// fixed number of args implies a static signature
-	if d.numberOfArgs > 0 {
-		args := make([]string, 0, d.numberOfArgs)
-		for i := 1; i <= d.numberOfArgs; i++ {
+	if d.NumberOfArgs > 0 {
+		args := make([]string, 0, d.NumberOfArgs)
+		for i := 1; i <= d.NumberOfArgs; i++ {
 			args = append(args, "arg"+strconv.Itoa(i))
 		}
 		argsPart := fmt.Sprintf("(%s interface{}) interface{}", strings.Join(args, ", "))
-		signature := d.name + argsPart
+		signature := d.Name + argsPart
 		return []string{signature}
 	}
 
 	// multi signatures
 	var signatures []string
-	for _, signature := range d.signatures {
-		signatures = append(signatures, d.name+signature)
+	for _, signature := range d.Signatures {
+		signatures = append(signatures, d.Name+signature)
 	}
 
 	return signatures
 }
 
 func (d dslFunction) Exec(args ...interface{}) (interface{}, error) {
 	// fixed number of args implies the possibility to perform matching between the expected number of args and the ones provided
-	if d.numberOfArgs > 0 {
-		if len(args) != d.numberOfArgs {
-			signatures := d.Signatures()
+	if d.NumberOfArgs > 0 {
+		if len(args) != d.NumberOfArgs {
+			signatures := d.GetSignatures()
 			if len(signatures) > 0 {
 				return nil, fmt.Errorf("%w. correct method signature %q", ErrInvalidDslFunction, signatures[0])
 			}
 			return nil, ErrInvalidDslFunction
 		}
 	}
 
-	return d.expressionFunction(args...)
+	// result cache check
```

Comment:
```
if !d.IsCacheable {
		return d.ExpressionFunction(args...)
	}

	functionHash := d.hash()
	if result, err := resultCache.Get(functionHash); err == nil {
		return result, nil
	}

	result, err := d.ExpressionFunction(args...)
	if err == nil {
		_ = resultCache.Set(functionHash, result)
	}
	return result, err
```

By doing this, we can avoid the need to check twice if the DSL expression can be cached.

Reply:
true that, but on the other side we need to repeat twice `d.ExpressionFunction(args...)`, really don't know what is more readable. What do you think?